### DISCUSSION TITLE
dokku security model

### DISCUSSION THREAD

**Discussion by cburkins** [dokku/dokku#4927] [Q&A] [answered]
New to dokku, and loving it!  It's a wonderful hosting solution.

I'm trying to understand the dokku security model. Namely, I've got a team of people who have permission to log into my dokku host.  Once they're logged in, and sitting at the CLI, I'm trying to figure out who's allowed to run "/usr/bin/dokku" on the dokku host command line.  Is there a way to limit and control that ?

**Comment by Cellane** [comment]
I think once the user is a member of `docker` group and can elevate privileges to root, they can control every aspect of `dokku` once they have SSH access to the host. If you’re looking for per-command and per-app access restrictions, https://github.com/dokku-community/dokku-acl would be the way – though even with this plugin installed and configured, once users have SSH access to the Dokku host, they can bypass the configuration. The intended use for the ACL plugin is via local `dokku` apps, as the documentation mentions.

Hope the above info can get you started!

**Reply by cburkins** [reply]
Not looking for app-level granular control, but thanks for the suggestion ! 

Our current design allows most of my team to log into this machine, and I'm looking to prevent most of them from being able to start/stop the dokku apps, and see all the associated dokku environment variables, which have username/passwords for associated services like the database.  

In the meantime, I had a colleague log into the machine, and try to run `/usr/bin/dokku`.  Command failed, because it appears that dokku requires using sudo to get root access, and the colleague didn't have that based on /etc/sudoers restrictions on the machine.  So I guess that's one way to control who can use "dokku", as they also need proper sudo rights to that comment.

I was also thinking that the "dokku" group might play into access control, but then I noticed that I'm not a member of the dokku user group, and I'm running dokku commands just fine.

Didn't think about the docker group, let me go double-check that.

**Reply by cburkins** [reply]
It appears that "dokku" is a member of the "docker" group.   But that's it.  I'm not a member of "dokku" nor "dockker".  So I guess control over the dokku capability on the machine is entirely premised on being able to run /usr/bin/dokku, which in turn needs sudo access ?

**Comment by josegonzalez** [comment, marked as answer]
The `dokku` command will [re-execute itself](https://github.com/dokku/dokku/blob/dacbb0f2dcfca1e736b989e99b77b35a0dd1d369/dokku#L92-L99) as the `dokku` user via `sudo` if the executing user is not `dokku`. This allows folks who are ssh'd onto the box that are either `root` or have `sudo` access to the dokku binary (via `sudoers` for instance) to run `dokku` commands. The typical linux permission model applies there.

For remote access via `ssh dokku@$HOST dokku:command here`, the user's ssh key is what is controlling remote access (those are managed via the `ssh-keys` plugin. There is no extra control over this, though you can control which subcommands a particular user has access to via the aforementioned [dokku-acl](https://github.com/dokku-community/dokku-acl) plugin (it controls not only access to apps but also access to specific commands, such as `config:export` or whatever).

The `dokku` user is given access to the `docker` group via [the postinst script](https://github.com/dokku/dokku/blob/master/debian/postinst#L29) in debian packaging (and similar for yum). Because it can execute docker commands, the biggest potential security risk is access to the `docker-options` plugin, as well as doing arbitrary storage mounts (both of which can be escalated out of the container to do whatever, such as mount `/etc/passwd`).

Since Dokku is traditionally run for personal setups and you'd need ssh access to even perform anything on it, the above are _fine_ normally, but also a reason why I don't expose the `docker` command as is via a plugin (arbitrary docker commands being incapable of being locked down). This is also the reason why we don't run scripts for much on the host, preferring to run any user code within a container.

With the release of [Dokku Pro](https://dokku.com/docs/enterprise/pro/), I am leaning on including general team management (and therefore more granular access control over apps and commands) with that product. I don't think this is a need for the general user (students, people running a couple apps for themselves, self-hosters all fall in this bucket) and it's complex enough that I'd rather not do this as part of the general offering.

Hope that answers your questions.

**Reply by cburkins** [reply]
This is very help, thanks @josegonzalez .   I now understand my two choices, namely:
1. Control who can sudo to dokku account via /etc/sudoers
2. Control who's ssh keys are registered via ssh-keys plugin (I'll need to practice a bit more with this one to get the hang of it)

I think these controls are quite adequate, thanks for the clarification !

-Chad
